File: ruby/binary-search-tree/binary_search_tree.rb
class Bst
  attr_accessor :data, :left, :right

  def initialize(data)
    @data = data
  end

  def insert(new_data)
    if new_data <= data
      if left
        left.insert(new_data)
      else
        @left = Bst.new(new_data)
      end
    else
      if right
        right.insert(new_data)
      else
        @right = Bst.new(new_data)
      end
    end
  end

  def each(&block)
    left && left.each(&block)
    block.call(data)
    right && right.each(&block)
  end
end

File: javascript/binary-search-tree/binary-search-tree.js
function Bst(data) {
  this.data = data;
  this.right = null;
  this.left = null;
}

Bst.prototype.insert = function(data) {
  if (data <= this.data) {
    if (this.left) {
      this.left.insert(data);
    } else {
      this.left = new Bst(data);
    }
  } else if (data > this.data) {
    if (this.right) {
      this.right.insert(data);
    } else {
      this.right = new Bst(data);
    }
  }
};

Bst.prototype.each = function(fx) {
  if (this.left) {
    this.left.each(fx);
  }
  fx(this.data);
  if (this.right) {
    this.right.each(fx);
  }
};

module.exports = Bst;
